### PR TITLE
introduce multipart/mixed support

### PR DESCRIPTION
**Pull request by eyalfa**
introduce `multipart/mixed` support according to [rfc-1341](https://www.w3.org/Protocols/rfc1341/7_2_Multipart.html)

* still WIP at this point

### CODE REVIEW

**Comment by eyalfa** [comment]
@987Nabil I'd appreciate your feedback on this

File: zio-http/jvm/src/test/scala/zio/http/multipart/mixed/MixedSpec.scala [hunk 1]
```diff
@@ -0,0 +1,163 @@
...
+    suiteAll("empty") {
+      val empty = Mixed.fromParts(ZStream.empty, Boundary(defaultSep))
+
+      test("has no parts") {
+        empty.parts.runCollect.map { collected =>
+          zio.test.assert(collected)(Assertion.isEmpty)
+        }
+      }
+
+      test("emoty no preamble, no epilogue") {
```
Comment (987Nabil) [inline zio-http/jvm/src/test/scala/zio/http/multipart/mixed/MixedSpec.scala]: empty?

File: zio-http/shared/src/main/scala/zio/http/multipart/mixed/Mixed.scala [hunk 2]
```diff
@@ -0,0 +1,306 @@
...
+import zio.{Chunk, ZIO, ZNothing}
+
+import zio.stream.{ZChannel, ZPipeline, ZStream}
+
+import zio.http._
+import zio.http.endpoint.openapi.JsonSchema.StringFormat.UUID
+import zio.http.internal.FormAST
+import zio.http.multipart.mixed.Mixed.Parser
+
+case class Mixed(source: ZStream[Any, Throwable, Byte], boundary: Boundary, bufferSize: Int = 8192) {
```
Comment (987Nabil) [inline zio-http/shared/src/main/scala/zio/http/multipart/mixed/Mixed.scala]: All case classes should be final

File: zio-http/shared/src/main/scala/zio/http/multipart/mixed/Mixed.scala [hunk 3]
```diff
@@ -0,0 +1,306 @@
...
+      case (Some(mt), None)           =>
+        base.contentType(mt)
+      case _                          =>
+        base
+    }
+  }
+}
+
+object Mixed {
+  val crlf = Chunk[Byte]('\r', '\n')
```
Comment (987Nabil) [inline zio-http/shared/src/main/scala/zio/http/multipart/mixed/Mixed.scala]: make things private that are for internal use only

**Comment by eyalfa** [comment]
@987Nabil thanks for your review,
I've applied the requested fixes plus simplified the part content parsing, making it tail-recursive and stack safe while at it.

I'd appreciate further review + suggested tests and other additions (such as adding stuff to the `Body` interface or endpoint API)

File: zio-http/shared/src/main/scala/zio/http/multipart/mixed/Mixed.scala [hunk 4]
```diff
@@ -0,0 +1,304 @@
...
+  def parts: ZStream[Any, Throwable, Part] = ZStream
+    .unwrapScoped[Any] {
+      source.toChannel.toPull.map { pull =>
+        new Parser(boundary, pull, bufferSize).result
+      }
+    }
+
+}
+
+case class Part(headers: Headers, bytes: ZStream[Any, Throwable, Byte]) {
```
Comment (987Nabil) [inline zio-http/shared/src/main/scala/zio/http/multipart/mixed/Mixed.scala]: final

File: zio-http/shared/src/main/scala/zio/http/multipart/mixed/Mixed.scala [hunk 5]
```diff
@@ -0,0 +1,304 @@
+package zio.http.multipart.mixed
```
Comment (987Nabil) [inline zio-http/shared/src/main/scala/zio/http/multipart/mixed/Mixed.scala]: zio.http

File: zio-http/shared/src/main/scala/zio/http/multipart/mixed/Mixed.scala [hunk 6]
```diff
@@ -0,0 +1,304 @@
...
+      case _                          =>
+        base
+    }
+  }
+}
+
+object Mixed {
+  private val crlf = Chunk[Byte]('\r', '\n')
+
+  class Parser(boundary: Boundary, pull: ZIO[Any, Throwable, Either[Any, Chunk[Byte]]], bufferSize: Int) {
```
Comment (987Nabil) [inline zio-http/shared/src/main/scala/zio/http/multipart/mixed/Mixed.scala]: private[http] final

**Review by 987Nabil — CHANGES_REQUESTED** [review]
You can add to the body trait `asMultiparMixed` just like it has `asMultipartForm`.
If you don't need it yourself, I'd not touch the endpoint API for now.
If of course you would like to do it, feel free. But for me it's okay to do this later and merge this first.

File: zio-http/shared/src/main/scala/zio/http/multipart/mixed/Mixed.scala (hunk 5 above)
Comment (eyalfa) [reply zio-http/shared/src/main/scala/zio/http/multipart/mixed/Mixed.scala]: do u mean move it to the `zio.http` package, in this case I guess it should be renamed to `MultipartMixed`
right?

**Review by jdegoes — APPROVED** [review]

**Review by jdegoes — APPROVED** [review]

**Pull request merged**